openvdb/openvdb/points/PointStatistics.h

```diff
@@ -0,0 +1,609 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+//
+/// @author Nick Avramoussis
+///
+/// @file PointStatistics.h
+///
+/// @brief Functions to perform multi threaded reductions and analysis of
+///   arbitrary point attribute types. Each function imposes various
+///   requirements on the point ValueType (such as expected operators) and
+///   supports arbitrary point filters.
+///
+
+#ifndef OPENVEB_POINTS_STATISTICS_HAS_BEEN_INCLUDED
+#define OPENVEB_POINTS_STATISTICS_HAS_BEEN_INCLUDED
+
+#include "PointDataGrid.h"
+
+#include "../openvdb.h"
+#include "../Types.h"
+#include "../math/Math.h"
+#include "../tree/LeafManager.h"
+
+#include <tbb/parallel_reduce.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Evaluates the minimum and maximum active values of a point attribute.
+/// @details  This function sets the arguments min and max to the minimum and
+///   and maximum values of a given point attributes. The ValueType of the
+///   attribute must be copy constructible and support less than and greater
+///   than operators. The ValueType does not need to support zero initialization
+///   or define its own numerical limits. This method will throw only if the
+///   templated ValueType does not match the given attribute. This method will
+///   return true if min and max have been set, false otherwise. The function is
+///   deterministic.
+/// @note  The value type of the min/max calculations must match the value type
+///   of the attribute. For vectors and matrices, this results in per component
+///   comparisons. See evalExtents for magnitudes or more custom control.
```

Comment:
By default, `evalMinMax` works per component. There is an internal function `evalExtents` which I have not exposed, but see the unit tests for an example: https://github.com/AcademySoftwareFoundation/openvdb/blob/eeddd9b0b3eb4028f3616825406765b4e4109a59/openvdb/openvdb/unittest/TestPointStatistics.cc#L739

It allows for custom results to be produces for things like magnitudes